addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/CachedPollyTTSCloudImpl.java

```diff
@@ -0,0 +1,169 @@
+            purgeAgedFiles();
+            return audioFileInCache;
+        }
+
+        // if not in cache, get audio data and put to cache
+        try (InputStream is = getTextToSpeech(text, label, audioFormat);
+                FileOutputStream fos = new FileOutputStream(audioFileInCache);) {
+            copyStream(is, fos);
+            // write text to file for transparency too
+            // this allows to know which contents is in which audio file
+            File txtFileInCache = new File(cacheFolder, fileNameInCache + ".txt");
+            writeText(txtFileInCache, text);
+            // return from cache
+            return audioFileInCache;
+        } catch (IOException ex) {
+            logger.warn("Could not write {} to cache, return null", audioFileInCache, ex);
+            return null;
+        }
+    }
+
+    /**
+     * Gets a unique filename for a give text, by creating a MD5 hash of it. It
+     * will be preceded by the voice label.
+     *
+     * Sample: "Robert_00a2653ac5f77063bc4ea2fee87318d3"
+     */
+    private String getUniqueFilenameForText(String text, String label) {
+        byte[] bytesOfMessage;
+        MessageDigest md;
+        try {
+            bytesOfMessage = text.getBytes("UTF-8");
+        } catch (UnsupportedEncodingException ex) {
+            logger.error("Could not create MD5 hash for '{}'", text, ex);
+            return null;
+        }
+        try {
+            md = MessageDigest.getInstance("MD5");
+        } catch (NoSuchAlgorithmException ex) {
+            logger.error("Could not create MD5 hash for '{}'", text, ex);
+            return null;
+        }
+        byte[] md5Hash = md.digest(bytesOfMessage);
+        BigInteger bigInt = new BigInteger(1, md5Hash);
+        String hashtext = bigInt.toString(16);
+        // Now we need to zero pad it if you actually want the full 32
+        // chars.
+        while (hashtext.length() < 32) {
+            hashtext = "0" + hashtext;
+        }
+        String fileName = label + "_" + hashtext;
+        return fileName;
+    }
+
+    // helper methods
+
+    private void copyStream(InputStream inputStream, OutputStream outputStream) throws IOException {
+        byte[] bytes = new byte[READ_BUFFER_SIZE];
+        int read = inputStream.read(bytes, 0, READ_BUFFER_SIZE);
+        while (read > 0) {
+            outputStream.write(bytes, 0, read);
+            read = inputStream.read(bytes, 0, READ_BUFFER_SIZE);
+        }
+    }
+
+    private void writeText(File file, String text) throws IOException {
+        try (OutputStream outputStream = new FileOutputStream(file)) {
+            outputStream.write(text.getBytes("UTF-8"));
+        }
+    }
+
+    private void updateTimeStamp(File file) throws IOException {
+        // update use date for cache management
+        file.setLastModified(System.currentTimeMillis());
+    }
+
+    private void purgeAgedFiles() throws IOException {
```

Comment:
Nice. I will consider this in my approach for https://github.com/eclipse/smarthome/issues/6041.